### PR TITLE
server: return FailedPrecondition when talking to decom node

### LINKED ISSUE #65589: roachtest: decommission/randomized failed [missing event]
roachtest.decommission/randomized [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3007632&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3007632&tab=artifacts#/decommission/randomized) on master @ [73154b35a3f679861a102b7cfcbeaa9bfc1aa10f](https://github.com/cockroachdb/cockroach/commits/73154b35a3f679861a102b7cfcbeaa9bfc1aa10f):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/decommission/randomized/run_1
	decommission.go:856,retry.go:197,decommission.go:802,decommission.go:61,test_runner.go:792: unexpected diff(matrix, expMatrix):
		[[][]string[19] != [][]string[20]]
		[[node_decommissioning] [node_recommissioned] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_decommissioning] [node_decommissioning] [node_decommissioned] [node_decommissioned] [node_decommissioned]]
		vs.
		[[node_decommissioning] [node_recommissioned] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_decommissioning] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_recommissioned] [node_decommissioning] [node_decommissioning] [node_decommissioned] [node_decommissioned] [node_decommissioning] [node_decommissioned]]
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh decommission/randomized
```
</p>
</p>
</details>
/cc cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*decommission/randomized.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by aliher1911**
Previously, when server detected that outgoing call would go to a
decommissioned node it would shortcut with codes.PermissionDenied
to skip attempt. That would result in retry loop immediately aborting
without trying another replica or refreshing leaseholder. One of the
consequences of that is that writes to system.eventlog will lose
events if trying to reach decommissioned node.
To address that, server will return codes.FailedPrecondition which
will result in this replica being evicted and call retried.

Release note: None

Fixes #65589

### CODE REVIEW

**Review by erikgrinaker — COMMENTED** [review]
Looks good! Can't think of anything else we need to handle. Except maybe check that it doesn't break anything for clients when connecting to decommissioned nodes (e.g. the `cockroach` CLI, both SQL and admin RPCs), although I don't _think_ it should affect them.

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-:-MbmOeVBCWNVagaS22-z:b-64l0eh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1919 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbm8FxF5Ttz16VgcPK2:-Mbm8FxG--WHsGwaLOCl:b7v9u50) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1919)):*
> ```Go
> 			// in-between.
> 			//
> 			if withCommit && !grpcutil.RequestDidNotStart(err) && !grpcutil.IsFailedPreconditionError(err) {
> ```

I'd argue that `RequestDidNotStart` should return `true` for `FailedPrecondition` (if we take "precondition" to mean "pre-evaluation" for the entire request). In this specific case it seems to apply, at least, since what we care about here is whether any writes could have gone through despite the error.
___
*[pkg/rpc/context.go, line 1062 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmFoD-6og3CRPBqtKO:-MbmFoD-6og3CRPBqtKP:b-3sp38y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/rpc/context.go#L1062)):*
> ```Go
> 				ctx.masterCtx, "rpc.Context: grpc heartbeat", func(masterCtx context.Context) {
> 					err := ctx.runHeartbeat(conn, target, redialChan)
> 					if err != nil && !grpcutil.IsClosedConnection(err) && !grpcutil.IsAuthError(err) && !grpcutil.IsFailedPreconditionError(err) {
> ```

nit: Break lines at 100 characters. I don't really mind, but I know others do (surprised the linter isn't saying anything).
___
*[pkg/server/connectivity_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmN9AGEroEzP0R_vDj:-MbmN9AGEroEzP0R_vDk:bifappq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/connectivity_test.go#L394)):*
> ```Go
> 		// Trying to scan the scratch range from the decommissioned node should
> 		// now result in a permission denied error.
> 		_, err := decomSrv.DB().Scan(ctx, scratchKey, keys.MaxKey, 1)
> ```

Let's try to do a scan the other way as well (from clusterSrv to decomSrv), to test that the operation doesn't hang because of infinite retries.
___
*[pkg/server/server.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmB-LFCZlSlyxnYae1:-MbmB-LFCZlSlyxnYae2:b-juonn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/server.go#L286)):*
> ```Go
> 
> 	nodeTombStorage := &nodeTombstoneStorage{engs: engines}
> 	checkPingFor := func(ctx context.Context, nodeID roachpb.NodeID, c codes.Code) error {
> ```

This parameter could use a more descriptive name.
___
*[pkg/util/grpcutil/grpc_util.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmBJGs07LKW-9fVjPL:-MbmBJGs07LKW-9fVjPM:bqkcq6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/util/grpcutil/grpc_util.go#L108)):*
> ```Go
> // IsFailedPreconditionError returns true if err's Cause is an error produced by
> // gRPC due to failed precondition. This error is used internally to shortcut
> // known unavailable nodes and is used by kv retry loop. When outside loop its
> ```

I'd be even more specific here and say that it's returned for decommissioned peers -- we want to retry unavailable nodes, because they could become available again. Maybe specify which KV retry loop as well and that we want to retry range-addressed requests, since there's retry loops all over the place. :)

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[pkg/server/connectivity_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmN9AGEroEzP0R_vDj:-MbpAggH8h-qZ3kAkP8a:bio2gks) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/connectivity_test.go#L394)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Let's try to do a scan the other way as well (from clusterSrv to decomSrv), to test that the operation doesn't hang because of infinite retries.
</blockquote></details>

Hm, maybe this doesn't make sense; if we put a range solely on the decommissioned node then either that range would have been moved off of the node first, or the request would indeed hang (because we're specifically retrying these requests in the DistSender).

I think we should have a test for the scenario we're trying to solve here though, but I'm not sure how to set ip up. Maybe something like setting up a range across three nodes, moving the leaseholder to the node we want to decommission, quiesce the range, shut the node down, decommission it, and then try to run a scan over it -- this should eventually succeed once the leaseholder is moved. It's a bit involved though. Any better ideas @tbg?

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[pkg/server/connectivity_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmN9AGEroEzP0R_vDj:-MbpC1b6C2ikhhW3dsPL:bm0ls3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/connectivity_test.go#L394)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Hm, maybe this doesn't make sense; if we put a range solely on the decommissioned node then either that range would have been moved off of the node first, or the request would indeed hang (because we're specifically retrying these requests in the DistSender).

I think we should have a test for the scenario we're trying to solve here though, but I'm not sure how to set ip up. Maybe something like setting up a range across three nodes, moving the leaseholder to the node we want to decommission, quiesce the range, shut the node down, decommission it, and then try to run a scan over it -- this should eventually succeed once the leaseholder is moved. It's a bit involved though. Any better ideas @tbg?
</blockquote></details>

Maybe it'd be easier to just stop the node and set the decommissioned status for the node directly in KV (without actually going through the whole decommission process).

File: pkg/util/grpcutil/grpc_util.go [hunk 1]
```diff
@@ -103,6 +103,20 @@ func IsAuthError(err error) bool {
 	return false
 }
 
+// IsFailedPreconditionError returns true if err's Cause is an error produced by
+// gRPC due to failed precondition. This error is used internally to shortcut
+// known unavailable nodes and is used by kv retry loop. When outside loop its
```
Comment (tbg) [inline pkg/util/grpcutil/grpc_util.go:108]: both "known unavailable nodes" "is used by kv retry loop" is ambiguous, can you be more specific? For example

This error is returned when trying to connect to nodes that are decommissioned (and thus will never be reachable again, at least under their old NodeID). The recipient of this error will want to handle this error by either contacting a different node (DistSender does this) when possible, or by returning the error to the client (handling it similar to an AuthError).

File: pkg/kv/kvclient/kvcoord/dist_sender.go [hunk 2]
```diff
@@ -1916,7 +1916,7 @@ func (ds *DistSender) sendToReplicas(
 			// evaluating twice, overwriting another unrelated write that fell
 			// in-between.
 			//
-			if withCommit && !grpcutil.RequestDidNotStart(err) {
+			if withCommit && !grpcutil.RequestDidNotStart(err) && !grpcutil.IsFailedPreconditionError(err) {
```
Comment (tbg) [inline pkg/kv/kvclient/kvcoord/dist_sender.go:1919]: shouldn't `RequestDidNotStart` be taught to recognize both AuthError and FailedPreconditionError as guaranteeing that the request didn't start?

**Review by tbg — COMMENTED** [review]
I think somehow we have to make `IsAuthError` and `IsFailedPreconditionError` and `RequestDidNotStart` play together better. In all cases the caller is asking something about the request. They're either asking "should I try again?" or "is there any chance this actually got executed despite my getting an error"? (And maybe something else I'm not aware of). Rather than sprinkling `if !A && !b { ` around, can we clean this up by giving the callers methods that describe what they are actually trying to find out?

Btw I went through the callers of AuthError and I think there are some that we need to update. For example https://github.com/cockroachdb/cockroach/blob/65c30a386376b0f8a6c10a77722cba9c399b57b3/pkg/ccl/kvccl/kvtenantccl/connector.go#L332-L335 should propagate (not sure that we'll really be able to hit the failed precondition here since that code is on the SQL tenant, but still)

Then this

https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1864-L1870

this

https://github.com/cockroachdb/cockroach/blob/0cea9dc28cf5b7855be87b6b38d55d191f1b963f/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L262-L265

this

https://github.com/cockroachdb/cockroach/blob/dd325f0cadb0d2c8059209ff13c9ea2043451740/pkg/kv/kvserver/liveness/liveness.go#L338-L340

and the list continues, basically you need to audit all of the callers (I like the "Call Hierarchy" feature in Goland for this).

**Comment by tbg** [comment]
`IsConnectionRejected(err)` could be a good name here - it's true if the error proves that we are, for whatever reason, not allowed to connect (either because of an auth error, or because the recipient is decommissioned). This would also feed into `RequestDidNotStart`.

**Review by aliher1911 — COMMENTED** [review]
I was going through the call sites where we check auth error to abort early and what happening is that we only have two places where it makes sense to handle. Outside of those places we should never get FailedPrecondition as we will keep iterating until we exhaust replicas or hit some other error. I agree about replacing isauterror with a better name, because callers don't care about type of error, they only care if they should bail in all those cases or most of them. Let me try to change that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-:-Mbq0dXCDMb840gdyWs1:b-9nbbv4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

**Review by aliher1911 — COMMENTED** [review]
I think connector.go should retry, it goes through the nodes in a loop and FailedPredondition should evict node from its list as other retryable errors. This is actually second place I mentioned where retry makes sense. And then ambigous error is handled fine, but I'll add precondition check to request didn't start as advised.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-:-Mbqm7_sBQEGHGUv3DVa:b-rdyvaz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911)

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1919 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbgyR6XPwq320mwx25b-r1-1919:-Mbqt33R0R8UuaGpk3ar:b-icmqsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1919)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

shouldn't `RequestDidNotStart` be taught to recognize both AuthError and FailedPreconditionError as guaranteeing that the request didn't start?
</blockquote></details>

In practice it never reaches this branch, but for the sake for consistency it should.
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1919 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbm8FxF5Ttz16VgcPK2:-MbrGD_f25GJpcqytNMS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/kv/kvclient/kvcoord/dist_sender.go#L1919)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'd argue that `RequestDidNotStart` should return `true` for `FailedPrecondition` (if we take "precondition" to mean "pre-evaluation" for the entire request). In this specific case it seems to apply, at least, since what we care about here is whether any writes could have gone through despite the error.
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 1062 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmFoD-6og3CRPBqtKO:-MbquHVQ2oj6-UaopnRI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/rpc/context.go#L1062)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: Break lines at 100 characters. I don't really mind, but I know others do (surprised the linter isn't saying anything).
</blockquote></details>

Done.
___
*[pkg/server/connectivity_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmN9AGEroEzP0R_vDj:-MbuVmQW0NXI9VLVfFkM:b-uggq8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/connectivity_test.go#L394)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Maybe it'd be easier to just stop the node and set the decommissioned status for the node directly in KV (without actually going through the whole decommission process).
</blockquote></details>

While it is possible here I think this is a wrong place to test this change in functionality. It should be verified on dist sender level for replica eviction logic.
___
*[pkg/server/server.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmB-LFCZlSlyxnYae1:-MbquGDJE6QcW_EJ3COQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/server.go#L286)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

This parameter could use a more descriptive name.
</blockquote></details>

Done.
___
*[pkg/util/grpcutil/grpc_util.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmBJGs07LKW-9fVjPL:-MbuccknA63uGSTjL66K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/util/grpcutil/grpc_util.go#L108)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'd be even more specific here and say that it's returned for decommissioned peers -- we want to retry unavailable nodes, because they could become available again. Maybe specify which KV retry loop as well and that we want to retry range-addressed requests, since there's retry loops all over the place. :)
</blockquote></details>

Done.
___
*[pkg/util/grpcutil/grpc_util.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbgyR6XPwq320mwx25k-r1-108:-Mbucd1X3b7cfaiWjI1A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/util/grpcutil/grpc_util.go#L108)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

both "known unavailable nodes" "is used by kv retry loop" is ambiguous, can you be more specific? For example

This error is returned when trying to connect to nodes that are decommissioned (and thus will never be reachable again, at least under their old NodeID). The recipient of this error will want to handle this error by either contacting a different node (DistSender does this) when possible, or by returning the error to the client (handling it similar to an AuthError).
</blockquote></details>

Done.

**Review by erikgrinaker — COMMENTED** [review]
> Reviewed 7 of 8 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1329 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbzXIWY0QqWd6VFRdVD:-MbzXIWY0QqWd6VFRdVE:ba4ooj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/kv/kvclient/rangecache/range_cache.go#L1329)):*
> ```Go
> 	// For now, all errors are retryable except transport errors where request was
> 	// rejected.
> 	return !grpcutil.IsConnectionRejected(err)
> ```

I think we want to retry `FailedPrecondition` here. In principle, if a range lookup fails then we want to retry that on the new leaseholder for the range. However, this is a bit hard to reason about, because there's cache eviction tokens and recursive DistSender calls involved, so please double-check this.
___
*[pkg/server/connectivity_test.go, line 394 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbmN9AGEroEzP0R_vDj:-MbzZINW9Cu15XQe9ONZ:b7dd52u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2e855a8c538e2c569b6d5749588963cf25889e/pkg/server/connectivity_test.go#L394)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

While it is possible here I think this is a wrong place to test this change in functionality. It should be verified on dist sender level for replica eviction logic.
</blockquote></details>

I agree, a DistSender unit test seems more appropriate here. Might be worth adding one for the range cache lookup retry as well (see my other comment).
___
*[pkg/server/server.go, line 312 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbz_TN63WvVXCgPmBpS:-Mbz_TN63WvVXCgPmBpT:b-ahl4hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/server/server.go#L312)):*
> ```Go
> 		Settings:   cfg.Settings,
> 		OnOutgoingPing: func(req *rpc.PingRequest) error {
> 			return checkPingFor(ctx, req.TargetNodeID, codes.FailedPrecondition)
> ```

I think we should have a comment here explaining why we return different error codes for incoming and outgoing pings.

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1329 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbzXIWY0QqWd6VFRdVD:-Mc9OIBS49fS8_sppDps:b-bcpj08) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/kv/kvclient/rangecache/range_cache.go#L1329)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I think we want to retry `FailedPrecondition` here. In principle, if a range lookup fails then we want to retry that on the new leaseholder for the range. However, this is a bit hard to reason about, because there's cache eviction tokens and recursive DistSender calls involved, so please double-check this.
</blockquote></details>

There are 3 call sites for this.
RangeIterator.Seek - it would always rely on underlying implementation of either DistSender or tennant Client. Both of them provide retry internally to handle this case.
DistSender.sendPartialBatch - this method sits on top of sendToReplicas where we perform retry.
DistSender.partialRangeFeed - is handling errors from routing info which uses range cache sitting on top of the same dist sender so it also retries internally.

The comment on the method states that its intended use is on range lookup calls, since range lookup or its underlying components are responsible for retry, this should be sufficient imho.
___
*[pkg/server/server.go, line 312 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbz_TN63WvVXCgPmBpS:-Mc9UJplDbCpNBCHHJeQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/server/server.go#L312)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I think we should have a comment here explaining why we return different error codes for incoming and outgoing pings.
</blockquote></details>

Done.

**Review by erikgrinaker — COMMENTED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[pkg/kv/kvclient/rangecache/range_cache.go, line 1329 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-MbzXIWY0QqWd6VFRdVD:-McDXlUuAyEyQ6WW6dlZ:ba1w2y9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/kv/kvclient/rangecache/range_cache.go#L1329)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

There are 3 call sites for this.
RangeIterator.Seek - it would always rely on underlying implementation of either DistSender or tennant Client. Both of them provide retry internally to handle this case.
DistSender.sendPartialBatch - this method sits on top of sendToReplicas where we perform retry.
DistSender.partialRangeFeed - is handling errors from routing info which uses range cache sitting on top of the same dist sender so it also retries internally.

The comment on the method states that its intended use is on range lookup calls, since range lookup or its underlying components are responsible for retry, this should be sufficient imho.
</blockquote></details>

Makes sense, thanks for verifying.
___
*[pkg/server/server.go, line 312 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbz_TN63WvVXCgPmBpS:-McDXWTw76VeeJ6HvIBy:b2vu10s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/server/server.go#L312)):*
<details><summary><i>Previously, aliher1911 (Oleg) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

s/unavailable/decommissioned/

Also, maybe reword this to clarify that the error code is _only_ returned when the given node is decommissioned. The way it's written it sounds like it's always blocked/rejected. :)
___
*[pkg/util/grpcutil/grpc_util.go, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-McDYhAg712q-SnvDs1z:-McDYhAg712q-SnvDs2-:bonuegi) ([raw file](https://github.com/cockroachdb/cockroach/blob/84af023db9d5e19704496d8355df3ff002a14baa/pkg/util/grpcutil/grpc_util.go#L152)):*
> ```Go
> 		errors.HasType(err, (*netutil.InitialHeartbeatFailedError)(nil)) ||
> 		errors.Is(err, circuit.ErrBreakerOpen) ||
> 		IsFailedPreconditionError(err) {
> ```

Shouldn't this check IsConnectionRejectedError? Let's expand the test with this as well.

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

___
*[pkg/server/server.go, line 312 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-Mbz_TN63WvVXCgPmBpS:-McELMjx534vrM0MNbPU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e01790e7bfc2b9059e6798a2dbdfd04d0f9923/pkg/server/server.go#L312)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

s/unavailable/decommissioned/

Also, maybe reword this to clarify that the error code is _only_ returned when the given node is decommissioned. The way it's written it sounds like it's always blocked/rejected. :)
</blockquote></details>

Done.
___
*[pkg/util/grpcutil/grpc_util.go, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-McDYhAg712q-SnvDs1z:-McELvad6dOjgeKJK7JF:b-uw6b06) ([raw file](https://github.com/cockroachdb/cockroach/blob/84af023db9d5e19704496d8355df3ff002a14baa/pkg/util/grpcutil/grpc_util.go#L152)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Shouldn't this check IsConnectionRejectedError? Let's expand the test with this as well.
</blockquote></details>

That makes sense for consistency even if we should never reach here with auth errors. We need some better clarity on which layers handle which errors to make those trivial changes less intrusive to the whole kv codebase.

**Review by erikgrinaker — APPROVED** [review]
> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @tbg)

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 5 of 8 files at r2, 1 of 1 files at r3, 3 of 4 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[pkg/kv/kvserver/liveness/liveness.go, line 338 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-McEh7-aECJpCyxJKiYA:-McEh7-bBim1tkoq773i:bcoaqjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d11a42c1f268fb31725438b678e1885f6643bd20/pkg/kv/kvserver/liveness/liveness.go#L338)):*
> ```Go
> 				log.Infof(ctx, "attempting to set liveness draining status to %v: %v", drain, err)
> 			}
> 			if grpcutil.IsConnectionRejected(err) {
> ```

We can't get FailedPrecondition here, right? Because we are going through a DistSender. Just checking my understanding. If we got FailedPrecondition here we should retry it (but we will never).

**Review by aliher1911 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @tbg)

___
*[pkg/kv/kvserver/liveness/liveness.go, line 338 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66199#-McEh7-aECJpCyxJKiYA:-McEnf9p5yjyRwvoRGiW:b-wueb0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d11a42c1f268fb31725438b678e1885f6643bd20/pkg/kv/kvserver/liveness/liveness.go#L338)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We can't get FailedPrecondition here, right? Because we are going through a DistSender. Just checking my understanding. If we got FailedPrecondition here we should retry it (but we will never).
</blockquote></details>

It uses DB.Txn to commit the change so my understanding is that it is above the distsender. I think it is safe.

**Comment by erikgrinaker** [comment]
Think we may have encountered this in the wild:

https://github.com/cockroachlabs/support/issues/1028#issuecomment-862585630

In this PR, when we send a request to a decommissioned node and that request fails, do we invalidate/update the relevant caches such that we don't try using that replica again?

**Comment by aliher1911** [comment]
We do evict lease when this happens. Without this fix we just bail without touching the lease token.

**Comment by erikgrinaker** [comment]
Great, thanks. I'm guessing that also evicts the range descriptor from its cache?

**Comment by aliher1911** [comment]
bors r=erikgrinaker,tbg

**Comment by erikgrinaker** [comment]
How do we feel about backporting this to 21.1?

**Comment by aliher1911** [comment]
I think we need to brew it on master. I traced all the code paths where this error could percolate, but lets wait at least for weekly to run that maybe? I don't think it would be hard technically to backport.

**Comment by erikgrinaker** [comment]
@aliher1911 Mind prepping a 21.1 backport so that this is at least ready to go when we feel confident about it? DoorDash got hit pretty badly by this (blocked all changefeed creation) and would like a near-term fix.

**Comment by erikgrinaker** [comment]
@aliher1911 Mind prepping a 21.1 backport so that this is at least ready to go when we feel confident about it? A customer got hit pretty badly by this (blocked all changefeed creation) and would like a near-term fix.
